### DISCUSSION TITLE
Is get_dirs() in template_loader.py supposed to execute continuously every second?

### DISCUSSION THREAD

**Discussion by mbelyayev** [django-components/django-components#848] [Q&A] [answered]
I've added LOGGING setting to settings.py with "root" logging setting that has level DEBUG:

```
LOGGING = {
        "version": 1,
        "disable_existing_loggers": False,
        "handlers": {
            "console": {
                "class": "rich.logging.RichHandler",
                "rich_tracebacks": True,
            },
        },
        "root": {
            "handlers": ["console"],
            "level": "DEBUG",
            "propogate": False
        },
    }
```
After adding a logger at DEBUG level, I've noticed the following gets printed from `logger.debug` statements in `template_loader.py` continuously @ 1x/sec, which pollutes the console. _Sidenote: let's say I have apps `components`, `core`, and `accounts` in `django-project`_:

```
[12/13/24 12:26:47] 
DEBUG    Template loader will search for valid template dirs from template_loader.py:33 following options:                                                            
    - /Users/user/projects/django-project/components                                                                    
    - /Users/user/projects/django-project/core/static                                                                                                                         
    - /Users/user/projects/django-project/accounts/static                                                                                                               
DEBUG    Template loader matched following template dirs:         template_loader.py:92                                       
    - /Users/user/projects/django-project/core/static                                                                   
    - /Users/user/projects/django-project/accounts/static                                                               
    - /Users/user/projects/django-project/components

[12/13/24 12:26:48] 
DEBUG """
DEBUG """

[12/13/24 12:26:49]
DEBUG """
DEBUG """

...                                                           
```  

Is this logging behavior from `django-components` expected behavior? If so, could it be explained in some fairly simple terms? Thank you!!

**Comment by JuroOravec** [comment]
Which version of Django-components, Django, and Python are you running?

In the log entries, for the latter two entries, `[12/13/24 12:26:48]` and `[12/13/24 12:26:49]`, is it just `DEBUG """` like you've shown above, or did you modify that to imply that the latter two entries had the same content?

But I don't think is coming from django-components. When I run the demo project in `sampleproject` directory of this repo, I don't see that behavior.

Assuming you are getting messages that look like `Template loader will search for valid template` every second:
1. Could you get a trace stack so we can see what is triggering the `template_loader`?
2. Could you try to make a minimal standalone Django project to reproduce and isolate this?

**Reply by mbelyayev** [reply]
Thanks for the quick response! Django-components 0.77, Django 5.0.6, Python 3.12.3.

With regards to the latter two entries and `DEBUG """`, yes I meant to say they are the exact same as the first log entry.

To your second question, I will try to follow up soon with a minimal reproducible example of this behavior.

To your first question, I added `traceback.print_stack()` under `get_dirs()` and am getting the following:

*my initial suspicion is the behavior is caused by the django template loading/autoreloading, or my using `runserver_plus` command for https during development; alas unless root cause is easy to spot I need more time to investigate*
```
 File "/Users/user/projects/django-project/manage.py", line 22, in <module>
    main()
  File "/Users/user/projects/django-project/manage.py", line 19, in main
    execute_from_command_line(sys.argv)
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django_extensions/management/utils.py", line 62, in inner
    ret = func(self, *args, **kwargs)
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django_extensions/management/commands/runserver_plus.py", line 338, in handle
    self.inner_run(options)
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django_extensions/management/commands/runserver_plus.py", line 459, in inner_run
    run_simple(
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/werkzeug/serving.py", line 1106, in run_simple
    run_with_reloader(
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/werkzeug/_reloader.py", line 456, in run_with_reloader
    reloader.run()
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/werkzeug/_reloader.py", line 257, in run
    self.run_step()
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/werkzeug/_reloader.py", line 297, in run_step
    for name in _find_stat_paths(self.extra_files, self.exclude_patterns):
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django_extensions/management/commands/runserver_plus.py", line 103, in extra_files
    template_files = get_all_template_files()
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django_extensions/management/commands/runserver_plus.py", line 84, in get_all_template_files
    for template_dir in get_template_directories():
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/template/autoreload.py", line 25, in get_template_directories
    items.update(
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/template/autoreload.py", line 25, in <genexpr>
    items.update(
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/template/loaders/cached.py", line 23, in get_dirs
    yield from loader.get_dirs()
  File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django_components/template_loader.py", line 28, in get_dirs
    traceback.print_stack()
```

**Reply by JuroOravec** [reply]
From the stack it seems that:
1. You run the `runserver_plus` command, which sets up `Werkzeug` [here](https://github.com/django-extensions/django-extensions/blob/main/django_extensions/management/commands/runserver_plus.py#L92)
2. For some reason, `Werkzeug` reloads
3. When `Werkzeug` reloads, it accesses the `self.extra_files`
4. `django_extensions` defines the `self.extra_files` [here](https://github.com/django-extensions/django-extensions/blob/main/django_extensions/management/commands/runserver_plus.py#L103) as a getter, which internally runs `django.template.autoreload.get_template_directories()`
5. django-components' template loader is run as part of `get_template_directories()`

So the question is what is causing Werkzeug to reload.

Btw I recommend to update to latest django-components if possible, because the code for template loader has been refactored between 0.77 and now. So potentially there could have been a bug in 0.77.

**Reply by mbelyayev** [reply]
I looked into this some more and confirmed the `Werkzeug` reloading behavior is standard and to be expected when using `runserver_plus`; see [here](https://werkzeug.palletsprojects.com/en/stable/serving/#reloader).

Therefore, it's to be expected that this will get called in the stack trace upon every reload:
```
File "/Users/user/projects/django-project/.venv/lib/python3.12/site-packages/django/template/loaders/cached.py", line 23, in get_dirs
    yield from loader.get_dirs()
```
However, now my question becomes this:

Why does any `logger.debug()` statement in the `loaders` mentioned below in `TEMPLATES` get silenced, except for `django_components.template_loader.Loader`?
```
TEMPLATES = [{
             ...
            'OPTIONS': {
                ...
                'loaders':[('django.template.loaders.cached.Loader', [
                        'django.template.loaders.filesystem.Loader',
                        'django.template.loaders.app_directories.Loader',
                        'django_components.template_loader.Loader',
                    ]
                )],
                ...
            },
            ...
        }]
```

For instance if I add the following to each respective loader's `get_dirs` function, only that of the `django_components.template_loader.Loader` will output a debug log `'test log'` to the console. 
```
import logging
logger = logging.getLogger(__name__)
logger.debug('test log')
```
This is certainly what's causing the pollution I'm experiencing.

**Reply by mbelyayev** [reply, marked as answer]
I guess this has become more of a django logging question; why doesn't `"root"` logger in `LOGGING` log from django modules such as `django.template`? To log django modules, why do you have to specify `"loggers"` key like so in `LOGGING` when `"root"` is already setup at DEBUG level:
```
"loggers": {
            "django.template.loaders.app_directories": {
                "handlers": ["console"],
                "level": "DEBUG",
                "propagate": False,
            },
        },
```
_Edit:_ The answer is explained [here](https://stackoverflow.com/a/22336174/19228033). You learn something new everyday!

Feel free to close this discussion. No issues with `django-components`.

**Reply by JuroOravec** [reply]
Hm, I might be wrong, but I don't think Werkzeug should be reloading every second. E.g. Django watches filesystem changes, and reloads only when a file changes. I would expect Werkzeug to work similarly. Do you possibly run some codegen in your project, and is it possible that this is being triggered on every reload?

**Reply by mbelyayev** [reply]
Upon further reading it seems `runserver` or `runserver_plus` both use a reloading mechanism that poll the filesystem every second for changes. So it's not reloading every second; rather it's polling (and loading templates) every second. Once the reloader conducts a poll where a change is detected, it reloads the server.

**Reply by JuroOravec** [reply]
Interesting, didn't know that 👀
